spring-integration-websocket/src/test/java/org/springframework/integration/websocket/ClientWebSocketContainerTests.java

```diff
@@ -105,6 +113,75 @@ public void testClientWebSocketContainer() throws Exception {
 	}
 
 
+	@Test
+	public void testRecoveryAfterRestart() throws Exception {
+
+		final AtomicBoolean failure = new AtomicBoolean();
+
+		StandardWebSocketClient webSocketClient = new StandardWebSocketClient() {
+
+			@Override
+			protected ListenableFuture<WebSocketSession> doHandshakeInternal(WebSocketHandler webSocketHandler,
+					HttpHeaders headers, URI uri, List<String> protocols, List<WebSocketExtension> extensions,
+					Map<String, Object> attributes) {
+
+				ListenableFuture<WebSocketSession> future =
+						super.doHandshakeInternal(webSocketHandler, headers, uri, protocols, extensions,
+								attributes);
+				if (failure.get()) {
+					future.cancel(true);
+				}
+
+				return future;
+			}
+
+		};
+
+		ClientWebSocketContainer container =
+				new ClientWebSocketContainer(webSocketClient, server.getWsBaseUrl() + "/ws/websocket");
+
+
+		TestWebSocketListener messageListener = new TestWebSocketListener();
+		container.setMessageListener(messageListener);
+		container.setConnectionTimeout(30);
+
+		container.start();
+
+		WebSocketSession session = container.getSession(null);
+		assertNotNull(session);
+		assertTrue(session.isOpen());
+
+		SubProtocolWebSocketHandler webSocketHandler = server.getServerContext()
+				.getBean(SubProtocolWebSocketHandler.class);
+		webSocketHandler.stop();
+
+		assertTrue(messageListener.sessionEndedLatch.await(10, TimeUnit.SECONDS));
+		assertFalse(session.isOpen());
+
+		container.stop();
+
+		failure.set(true);
+
+		container.start();
+
+		try {
+			container.getSession(null);
+			fail("IllegalStateException is expected");
+		}
+		catch (Exception e) {
+			assertThat(e, instanceOf(IllegalStateException.class));
+			assertThat(e.getCause(), instanceOf(CancellationException.class));
+		}
+
+		failure.set(false);
+
+		container.start();
```

Comment:
See `container.getSession()` logic.
If we catch an `Exception`, we `stop()` the container.
Therefore before the next `getSession()` we should `start()` it manually.

That was always exactly premise but without the proper test and clean up for that `openConnectionException` property...

Right now we move it to stopped state after the failures. You should manually decide if it should be restarted or not.
Yes, we may consider scheduled recovery in the future.
Not sure if we should try to address it right now...